extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/telemetry/TracesForwardingWebSocketEndpoint.java

```diff
@@ -0,0 +1,92 @@
+package io.quarkus.websockets.next.runtime.telemetry;
+
+import static io.quarkus.websockets.next.runtime.telemetry.TelemetryConstants.CONNECTION_CLIENT_ATTR_KEY;
+import static io.quarkus.websockets.next.runtime.telemetry.TelemetryConstants.CONNECTION_ENDPOINT_ATTR_KEY;
+import static io.quarkus.websockets.next.runtime.telemetry.TelemetryConstants.CONNECTION_ID_ATTR_KEY;
+
+import java.util.function.Function;
+
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.semconv.UrlAttributes;
+import io.quarkus.websockets.next.WebSocketClientConnection;
+import io.quarkus.websockets.next.WebSocketConnection;
+import io.quarkus.websockets.next.runtime.WebSocketEndpoint;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+
+/**
+ * {@link WebSocketEndpoint} wrapper that produces OpenTelemetry spans for closed connection.
+ */
+final class TracesForwardingWebSocketEndpoint extends ForwardingWebSocketEndpoint {
+
+    /**
+     * Target ID represents either endpoint id or client id.
+     */
+    private final String targetIdKey;
+    private final String targetIdValue;
+    /**
+     * Span context for an HTTP request used to establish the WebSocket connection.
+     */
+    private final Tracer tracer;
+    private final String connectionId;
+    private final String path;
+    private final SpanContext onOpenSpanContext;
+    private final SpanKind spanKind;
+
+    TracesForwardingWebSocketEndpoint(WebSocketEndpoint delegate, Tracer tracer, WebSocketConnection connection,
+            SpanContext onOpenSpanContext, String path) {
+        super(delegate);
+        this.tracer = tracer;
+        this.onOpenSpanContext = onOpenSpanContext;
+        this.connectionId = connection.id();
+        this.targetIdKey = CONNECTION_ENDPOINT_ATTR_KEY;
+        this.targetIdValue = connection.endpointId();
+        this.path = path;
+        this.spanKind = SpanKind.SERVER;
+    }
+
+    TracesForwardingWebSocketEndpoint(WebSocketEndpoint delegate, Tracer tracer, WebSocketClientConnection connection,
```

Comment:
Hm, again, a matter of style but it's usually clearer to keep one logicless constructor and delegate from specific ones, e.g. something like:

```java
    TracesForwardingWebSocketEndpoint(WebSocketEndpoint delegate, Tracer tracer, WebSocketConnection connection, SpanContext onOpenSpanContext, String path) {
      this(delegate, tracer, connection.id(), CONNECTION_ENDPOINT_ATTR_KEY, connection.endpointId(), onOpenSpanContext, path, SpanKind.SERVER);
    }
   
    TracesForwardingWebSocketEndpoint(WebSocketEndpoint delegate, Tracer tracer, WebSocketClientConnection connection, SpanContext onOpenSpanContext, String path) {
      this(delegate, tracer, connection.id(), CONNECTION_CLIENT_ATTR_KEY, connection.clientId(), onOpenSpanContext, path, SpanKind.CLIENT);
    }

    private TracesForwardingWebSocketEndpoint(WebSocketEndpoint delegate, Tracer tracer, String connectionId, String targetIdKey, String targetIdValue, SpanContext onOpenSpanContext, String path, SpanKind spanKind) {
        super(delegate);
        this.tracer = tracer;
        this.onOpenSpanContext = onOpenSpanContext;
        this.connectionId = connectionId;
        this.targetIdKey = targetIdKey;
        this.targetIdValue =targetIdValue;
        this.path = path;
        this.spanKind = spanKind;
    }
```